File: DSAlgorithms/Leetcode/c++/arrays_hash/496-next_greater_element.cpp
vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
    std::stack<int> a_stack;
    std::vector<int> result(nums1.size(), -1);
    std::unordered_map<int, int> hash;
    for(int i = 0; i < nums1.size(); i++)
        hash[nums1[i]] = i;
    for(int i = 0; i < nums2.size(); i++)
    {
        int current = nums2[i];
        while(!a_stack.empty() && current > a_stack.top())
        {
            int val = a_stack.top();
            a_stack.pop();
            int index = hash[val];
            result[index] = current;
        }
        if(hash.count(current) > 0)
            a_stack.push(current);
    }
    return result;
}

File: DSAlgorithms/Leetcode/ts/src/arrays_hash/496-next_greater_element.ts
const nextGreaterElement = (nums1: number[], nums2: number[]): number[] =>{
    const a_stack: number[] = [];
    const result: number[] = Array(nums1.length).fill(-1);
    const a_hash = new Map<number, number>();
    for(let i = 0; i < nums1.length; i++)
        a_hash.set(nums1[i], i);
    for(let i = 0; i < nums2.length; i++)
    {
        let current = nums2[i];
        while(a_stack.length > 0 && current > a_stack[a_stack.length - 1])
        {
            let val = a_stack[a_stack.length - 1];
            a_stack.pop();
            let index = a_hash.get(val);
            result[index] = current;
        }
        if(a_hash.has(current))
            a_stack.push(current);
    }
    return result;
};
